addons/website/static/tests/tours/edit_link_popover.js

```diff
@@ -185,20 +187,21 @@ wTourUtils.registerWebsitePreviewTour('edit_link_popover', {
         run: function () {}, // it's a check
     },
     {
-        content: "Ensure that a click on the link popover link opens a new window",
+        content: "Ensure that a click on the link popover link opens a new window in edit mode",
         trigger: 'iframe .o_edit_menu_popover a.o_we_url_link[target="_blank"]',
         extra_trigger: 'iframe .o_edit_menu_popover a.o_we_full_url[target="_blank"]',
         run: function (actions) {
-            this.$anchor[0].addEventListener('click', (ev) => {
-                const originalPreventDefault = ev.preventDefault.bind(ev);
-                ev.preventDefault = () => {
-                    throw new Error(
-                        "The link popover should not be default prevented as to open a new tab"
-                    );
-                };
-                originalPreventDefault(); // We do not want to open a new tab in a tour
-            }, {once: true, capture: true});
+            patch(browser, 'window_open_action', {
+                open: (url) => {
+                    if (window.location.hostname === url.hostname && !url.pathname.startsWith('/@/')) {
+                        throw new Error(
+                            "The link popover should redirect to the backend link."
+                        );
+                    }
+                }
+            }, { pure: true }) // We do not want to open a link in a tour
             actions.click();
+            unpatch(browser,'window_open_action')
```

Comment:
Hello @detrouxdev @rdeodoo,
Here is the implementation of the way you suggested.

Before this, the test was focused on testing that a link should open in a new window. 
But IMO now, the test focus should be on testing that link should open in backend mode, no?

There are 3 cases we have,
 
1. open a link in a new window in a tour.
I have doubts about this, do we need to open a link during the tour?
If yes, we can do something like this by calling `window.open(URL)`.
![new_window](https://github.com/odoo/odoo/assets/125337508/78248c1c-2ff2-483c-9ea7-49c3d4aa4a5f)

2. Open a link in the same window in a tour.
Is it really will be good to change the actual work of function as the tour will be a guide for the user? like, when the user clicks on the link it will redirect to the new window & while in the tour, it won't. Is this desired behaviour? 
Also, it will create an issue when there are more tour steps we will have after this. ( as the page will be refreshed after this tour step )

3. Just test that a link URL is correct, don't open the link. ( which I have implemented currently )
In this, ensure that the link should open in edit mode without opening it. 

What are your thoughts on it?

Reply:
This is a test tour so there's no need to worry about the user! It's just a way for us to programatically test the UI.

I think each option has a valid positive to it.

1. This would not modify the behaviour and would ensure we're as close to reality as possible
2. We could technically redirect but I think this is not a good idea and could lead to race conditions.
3. IMO this is the best option, you check the URL and it's good enough for me.

I don't think you should patch and unpatch in the same step, I'm not 100% sure, but there's a risk that you unpatch while the event handler hasn't even been processed yet. Instead you could modify the DOM in the patched open method and check if the DOM is properly modified in the next step. 😄

Reply:
Ok!! then good to go with option 3.

But do we really need to force modify the DOM, as the link click method will not have any DOM modification inside?
Agree with you @detrouxdev  there can be a risk, but that can be handled by making the method `async` too, I guess. no?

Something like, 
![async_function](https://github.com/odoo/odoo/assets/125337508/fe2d1eb3-8e10-437f-a5d7-3ae4a3fd3f7b)

Reply:
> somehow await for the open method to be called

:thinking:  Ok @detrouxdev. 

Hope this is what we are expecting, 
- it will resolve the promise as soon as the open method is executed and allowed to unpatch.

```javascript
run: async function (actions) {
    patch(browser, 'window_open_action', {
        open: (url) => {
            if (window.location.hostname === url.hostname && !url.pathname.startsWith('/@/')) {
                throw new Error("The link popover should redirect to the backend link.");
            }
            this.resPromise("The link popover will open link in backend.") 
        }
    }, { pure: true }) // We do not want to open a link in a tour
    await new Promise((resolve) => {
        this.resPromise = resolve
        actions.click();
    });
    unpatch(browser,'window_open_action')
}

// ( due to having an issue with variable scoping, I have to take this.resPormise variable ) 
```
What do think about this?

Reply:
If it works, it's okay with me.
Though I would not put the resolve inside this, just declare an empty variable with
`let resPromise`; at the top of the run method. You should have access to it in both your functions.

Reply:
@divy-odoo Are you sure that the `run()` is really awaited? I always thought it wasn't.
IMO, the test will finish before executing (fully?) this run javascript code.
I would expect the promise success to set some attribute in the DOM and a next step to check that attribute.
So, if the tour moves to the next step while this JS is not really finished, the step will not succeed until the JS did what is expected and the test is correctly tested.

From what I'd imagine, the test is:
- Checking the trigger and extra trigger of your step, which succeed
- Starting to run the custom js `Run` of that step
- Directly checking the next step triggers, which if found will succeed regardless of the previous run js code status
  But in this case there is no step after that anyway, so I expect the tour to be considered successful and finish at the same time as the run is executed.

Reply:
I actually gave it a test, and it's how I expected:
```
Tour Manager is ready.  running_tour=test_unsplash_beacon
TIP.RUN IS A FUNCTION, GOIGN TO CALL THE FUNCTION
body found, starting 4s sleep
Tour test_unsplash_beacon: step 'Verify whether beacon was sent. (trigger: body)' succeeded
TIP.RUN IS A FUNCTION, GOIGN TO CALL THE FUNCTION
body found (2), starting 4s sleep
Tour test_unsplash_beacon: step 'Verify whether beacon was sent. (trigger: body)' succeeded
Tour test_unsplash_beacon succeeded
test successful
end sleep
end sleep
```

This is with the following code:
```js
tour.register("test_unsplash_beacon", {
    test: true,
    url: "/?test_unsplash_beacon",
}, [{
    content: "Verify whether beacon was sent.",
    trigger: 'body',
    run: async () => {
        console.log('body found, starting 4s sleep');
        await sleep(4000);
        console.log('end sleep');
    },
},
{
    content: "Verify whether beacon was sent.",
    trigger: 'body',
    run: async () => {
        console.log('body found (2), starting 4s sleep');
        await sleep(4000);
        console.log('end sleep');
    },
}]);
});
```

Reply:
Ah I spend too much time in master I forgot when they converted the tour util into the new "macros" ( [done in 16.3](https://github.com/odoo/odoo/commit/3a798039d6f200f8e28448ddb6a2d3c46654a203) )
In master with the same code:
```
Tour test_unsplash_beacon on step: 'Verify whether beacon was sent. (trigger: body)'
unsplash_beacon.js:31 body found, starting 4s sleep
unsplash_beacon.js:33 end sleep
tour_compilers.js:295 Tour test_unsplash_beacon on step: 'Verify whether beacon was sent. (trigger: body)'
unsplash_beacon.js:40 body found (2), starting 4s sleep
unsplash_beacon.js:42 end sleep
tour_service.js:215 test successful
```

Doesn't matter much here but just thought it was interesting to note.

Pretty sure you will have to modify the DOM in the patched open function and add a step then @divy-odoo, at least up to <16.3

Reply:
@detrouxdev Nice, thank you for the precision.
@divy-odoo Just add a class to the body, no need of a new element :+1:

Reply:
Hello @rdeodoo, The changes as been applied as per the discussion.